x/dymns/keeper/dym_name.go

```diff
@@ -0,0 +1,737 @@
+				continue
+			}
+
+			if config.Path != subName {
+				continue
+			}
+
+			return config.Value, true
+		}
+
+		return "", false
+	}
+
+	// first attempt to resolve, to see if the chain-id/alias is user-configured
+	var found bool
+	outputAddress, found = tryResolveFromConfig(chainIdOrAlias)
+	if found {
+		return
+	}
+
+	// end of first attempt
+
+	var resolvedToChainId string
+
+	if chainId, success := k.tryResolveChainIdOrAliasToChainId(ctx, chainIdOrAlias); success {
+		resolvedToChainId = chainId
+	} else {
+		// treat it as chain-id
+		resolvedToChainId = chainIdOrAlias
+	}
+
+	// second attempt to resolve
+	outputAddress, found = tryResolveFromConfig(resolvedToChainId)
+	if found {
+		return
+	}
+	// end of second attempt
+
+	// no more attempts to resolve from config
+
+	// try fallback
+
+	if subName != "" {
+		// no need to fallback
+		return
+	}
+
+	if resolvedToChainId == ctx.ChainID() {
+		outputAddress = dymName.Owner
+		return
+	}
+
+	resolveToAddress := dymName.Owner
+	for _, config := range dymnstypes.DymNameConfigs(dymName.Configs).DefaultNameConfigs() {
+		resolveToAddress = config.Value
+		break
+	}
+
+	// not the host chain, try to resolve if is a RollApp
+
+	isRollAppId := k.IsRollAppId(ctx, resolvedToChainId)
+	if !isRollAppId {
+		// fallback does not apply for non-RollApp
+		return
+	}
+
+	rollAppBech32Prefix, found := k.GetRollAppBech32Prefix(ctx, resolvedToChainId)
+	if !found {
+		// fallback does not apply for RollApp does not have this metadata
+
+		// TODO DymNS: write test cover this case
+		return
+	}
+
+	accAddr := sdk.MustAccAddressFromBech32(resolveToAddress)
+	rollAppBasedBech32Addr, convertErr := bech32.ConvertAndEncode(rollAppBech32Prefix, accAddr)
+	if convertErr != nil {
+		err = sdkerrors.ErrInvalidAddress.Wrapf(
+			"failed to convert '%s' to RollApp-based address: %v", resolveToAddress, convertErr,
+		)
+		return
+	}
+
+	outputAddress = rollAppBasedBech32Addr
+	return
+}
+
+func (k Keeper) tryResolveChainIdOrAliasToChainId(ctx sdk.Context, chainIdOrAlias string) (resolvedToChainId string, success bool) {
+	if chainIdOrAlias == ctx.ChainID() {
+		return chainIdOrAlias, true
+	}
+
+	chainsParams := k.ChainsParams(ctx)
+	if len(chainsParams.AliasesOfChainIds) > 0 {
+		for _, record := range chainsParams.AliasesOfChainIds {
+			if chainIdOrAlias == record.ChainId {
+				return record.ChainId, true
+			}
+
+			for _, alias := range record.Aliases {
+				if alias == chainIdOrAlias {
+					return record.ChainId, true
+				}
+			}
+		}
+	}
+
+	if isRollAppId := k.IsRollAppId(ctx, chainIdOrAlias); isRollAppId {
+		return chainIdOrAlias, true
+	}
+
+	if rollAppId, found := k.GetRollAppIdByAlias(ctx, chainIdOrAlias); found {
+		// TODO DymNS: require RollApp alias to use dymnsutils.IsValidAlias to validate the alias
+		return rollAppId, true
+	}
+
+	return
+}
+
+func ParseDymNameAddress(
+	dymNameAddress string,
+) (
+	subName string, dymName string, chainIdOrAlias string, err error,
+) {
```

Comment:
I'm not sure about pure regex but I wouldn't choose that way because:
- Regex to read and understand the rules.
- Can not provide details content about which part of the params is invalid, describe it in details so user can fix.

Reply:
code now have 100% docs, please read again ty